### PR TITLE
Add Login controller

### PR DESCRIPTION
**Pull request by elland**
Thinking about making this into a component we can reuse for other apps as well. 🤔

### CODE REVIEW

**Comment by 3lvis** [comment]
@Elland make a separate repo for it. Would help to keep things more abstract.

File: iOS/Source/LoginController.swift [hunk 1]
```diff
@@ -120,7 +120,25 @@ class LoginController: SweetTableController {
     }
 
     func didPressSignInButton() {
-        print("Sign in!")
+        let username = "bakkenbaeck"
+        let password = "ta2Tz*yYQLRxbVEJ3RBhqXD&t"
```
Comment (3lvis) [inline iOS/Source/LoginController.swift]: Please hide this one.

File: iOS/Source/LoginController.swift [hunk 2]
```diff
@@ -130,9 +130,7 @@ class LoginController: SweetTableController {
                 guard let json = try JSONSerialization.jsonObject(with: data, options: .allowFragments) as? [String: String] else { return }
                 if let result = json["result"] {
                     self.fetcher.userToken = "\(username):\(result)"
-                    DispatchQueue.main.async {
-                        let _ = self.navigationController?.popToRootViewController(animated: true)
-                    }
+                    self.dismiss(animated: true)
```
Comment (3lvis) [inline iOS/Source/LoginController.swift]: Dismiss goes to the controller that presented this. Or am I missing something?

Comment (elland) [reply iOS/Source/LoginController.swift]: Temporary

**Comment by elland** [comment]
What do you think? 🤔

File: iOS/Source/LoginController.swift [hunk 3]
```diff
@@ -128,12 +154,31 @@ class LoginController: SweetTableController {
...
+
+            self.didPressSignInButton()
+        }
+    }
+
     func didPressSignInButton() {
         let username = self.loginParameters.username ?? ""
         let password = self.loginParameters.password ?? ""
-        let url = URL(string: "https://\(username):\(password)@api.pinboard.in/v1/user/api_token?format=json")!
+        let base64Credentials = "\(username):\(password)".data(using: .utf8)!.base64EncodedString()
```
Comment (3lvis) [inline iOS/Source/LoginController.swift]: https://github.com/3lvis/Networking#http-basic

Comment (elland) [reply iOS/Source/LoginController.swift]: Done.

File: iOS/Source/LoginController.swift [hunk 4]
```diff
@@ -168,36 +168,21 @@ class LoginController: SweetTableController {
...
+        self.fetcher.login() { (JSON, error) in
+            if let error = error {
+                var message = error.localizedDescription
+                if error.code == 401 {
+                    message = "Username/password combination invalid!"
                 }
+                let alert = UIAlertController.dismissableAlert(title: "Authentation failed", message: message)
+                self.present(alert, animated: true)
+            } else if let result = JSON?["result"] {
+                self.fetcher.userToken = "\(username):\(result)"
```
Comment (3lvis) [inline iOS/Source/LoginController.swift]: This should be done inside fetcher. The Controller shouldn't have knowledge about this.

File: iOS/Source/LoginController.swift [hunk 5]
```diff
@@ -168,36 +168,21 @@ class LoginController: SweetTableController {
...
+        self.fetcher.networking.authenticate(username: username, password: password)
+        self.fetcher.login() { (JSON, error) in
+            if let error = error {
+                var message = error.localizedDescription
+                if error.code == 401 {
+                    message = "Username/password combination invalid!"
                 }
+                let alert = UIAlertController.dismissableAlert(title: "Authentation failed", message: message)
+                self.present(alert, animated: true)
+            } else if let result = JSON?["result"] {
```
Comment (3lvis) [inline iOS/Source/LoginController.swift]: I think this should be moved to Fetcher, the controller shouldn't know about this.

File: iOS/Source/LoginController.swift (hunk 4 above)
Comment (elland) [reply iOS/Source/LoginController.swift]: Agreeed. 👍

**Comment by elland** [comment]
Up-da-ted.

**Comment by 3lvis** [comment]
Is this done or still WIP?

**Comment by elland** [comment]
@3lvis thinking about this. I have some ideas of how to do the form parsing better but that can be a new PR.

**Comment by 3lvis** [comment]
Merge it!

**Pull request merged**